src/main/java/macrobase/analysis/contextualoutlier/ContextualOutlierDetector.java

```diff
@@ -1,93 +1,408 @@
 package macrobase.analysis.contextualoutlier;
 
+import java.text.NumberFormat;
 import java.util.ArrayList;
+import java.util.BitSet;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
-
+import java.util.Random;
+import java.util.StringJoiner;
+import java.util.concurrent.TimeUnit;
+import java.io.*;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.base.Stopwatch;
+
+import macrobase.analysis.classify.OutlierClassifier;
+import macrobase.analysis.classify.StaticThresholdClassifier;
+import macrobase.analysis.result.OutlierClassificationResult;
 import macrobase.analysis.stats.BatchTrainScore;
+import macrobase.analysis.summary.itemset.FPGrowthEmerging;
+import macrobase.analysis.summary.itemset.result.ItemsetResult;
 import macrobase.analysis.summary.result.DatumWithScore;
+import macrobase.analysis.transform.BatchScoreFeatureTransform;
+import macrobase.analysis.transform.FeatureTransform;
+import macrobase.conf.ConfigurationException;
+import macrobase.conf.MacroBaseConf;
+import macrobase.conf.MacroBaseDefaults;
 import macrobase.datamodel.Datum;
+import macrobase.ingest.DatumEncoder;
+import macrobase.util.IterUtils;
 
 public class ContextualOutlierDetector{
     private static final Logger log = LoggerFactory.getLogger(ContextualOutlierDetector.class);
 
     
-    private BatchTrainScore detector; //the detector used for every context
-    
     
+    private MacroBaseConf conf;
     private List<String> contextualDiscreteAttributes;
     private List<String> contextualDoubleAttributes;
     private int totalContextualDimensions;
     
     
+    Context globalContext;
+    
+    
     private double denseContextTau;
     private int numIntervals;
-    
+    private int maxPredicates;
+    private ContextPruningOptions contextPruningOptions;
+    private DatumEncoder encoder;
     //This is the outliers detected for every dense context
-    private Map<Context,BatchTrainScore.BatchResult> context2Outliers = new HashMap<Context,BatchTrainScore.BatchResult>();
+    //could've stored Context,OutlierDetector.BatchResult, but waste a lot of memory
+    private Map<Context,OutlierClassifier> context2OutlierClassifier = new HashMap<Context,OutlierClassifier>();
+    
+    private PrintWriter contextualOut;
+    
+    public ContextualOutlierDetector(MacroBaseConf conf){
```

Comment:
I have a BasicContextualBatchedPipeline that extends AbstractPipeline, not OneShotPipeline, for reasons explained next.